### PR TITLE
Make wahtson accesible through node require()

### PR DESCRIPTION
**Pull request by barnabwhy**
idk what to call this branch

### CODE REVIEW

File: src/bot.js [hunk 1]
```diff
@@ -15,41 +15,55 @@ const { version } = require('../package.json')
...
+                config.get('bot_token', async token => {
+                    await client.login(token)
+                    return true
+                })
+            })
+            .catch(err => {
+                console.error(chalk.red(`error: ${err}`))
+                event.emit('error', `error: ${err}`)
+                process.exit(1)
+            })
```
[the remark this answers is not in the document]
Comment (barnabwhy) [reply src/bot.js:48]: ok

File: src/bot.js [hunk 2]
```diff
@@ -15,41 +15,55 @@ const { version } = require('../package.json')
...
+const EventEmitter = require('events')
+const event = new EventEmitter()
+
+module.exports = {
+    start: () => {
+        config
+            .load()
+            .then(() =>
+                sqlite.open({
+                    filename: './database.sqlite',
```
[the remark this answers is not in the document]
Comment (nanaian) [reply src/bot.js:30]: Either a file path string or a sqlite database instance

File: wahtson.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+const chalk = require('chalk')
+const bot = require("./src/bot.js");
```
Comment (nanaian) [inline wahtson.js:2]: Inconsistent use of quotes here; use single quotes in this file like the rest of the project

Comment (nanaian) [reply wahtson.js:2]: & below also

Comment (nanaian) [reply wahtson.js:2]: Sorry for the pedantry :shipit:

File: src/bot.js [hunk 4]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
+
+            send = async (a, b) => {
+                this.emit(a, b)
+            }
+            module.exports.send = async (a, b) => {
+                this.emit(a, b)
+            }
+            send('info', ['test', 'cyan'])
+        }
+        async start() {
```
Comment (nanaian) [inline src/bot.js:37]: Space out these methods a bit
```suggestion
        }

        async start() {
```

File: src/bot.js [hunk 5]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
+                .then(async () => {
+                    config.get('bot_token', async token => {
+                        await client.login(token)
+                        return true
+                    })
+                })
+                .catch(err => {
+                    send('fatal_error', err)
+                    process.exit(1)
+                })
```
Comment (nanaian) [inline src/bot.js:60]: Don't exit the process! Just reject the promise; don't catch at all. You don't need `fatal_error`.

Comment (nanaian) [reply src/bot.js:60]: It would probably make sense to just do `await config.load(configPath)` etc. rather than `then`-chaining.

File: src/bot.js [hunk 6]
```diff
@@ -81,29 +102,42 @@ client.on('message', async msg => {
             const member = msg.member || (await guild.members.fetch(msg.author))
             if (!(await config.has('on_message'))) return
             await executeActionChain(await config.get('on_message'), {
+                event_call: 'on_message',
                 message: msg,
                 channel: msg.channel,
                 member: msg.member,
                 command: null,
+                limitLog: (await config.get('on_message')).limit_log,
```
Comment (nanaian) [inline src/bot.js:110]: What is this?

File: src/bot.js [hunk 7]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
-            return true
-        })
-    })
-    .catch(err => {
-        console.error(chalk.red(`error: ${err}`))
-        process.exit(1)
-    })
+const EventEmitter = require('events')
+
+let send
```
Comment (nanaian) [inline src/bot.js:20]: :-1: This should be a method on the class.

File: src/bot.js [hunk 8]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
+            super()
+            this.botOptions = botOptions || {}
+            this.version = version
+
+            send = async (a, b) => {
+                this.emit(a, b)
+            }
+            module.exports.send = async (a, b) => {
+                this.emit(a, b)
+            }
```
Comment (nanaian) [inline src/bot.js:34]: 👎 👎 👎

File: src/bot.js [hunk 9]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
+module.exports = {
+    Wahtson: class Bot extends EventEmitter {
+        constructor(botOptions) {
+            super()
+            this.botOptions = botOptions || {}
+            this.version = version
+
+            send = async (a, b) => {
+                this.emit(a, b)
+            }
```
Comment (nanaian) [inline src/bot.js:31]: Surely you can just use `this.emit(a, b)` where `send(a, b)` is used?

File: src/bot.js [hunk 10]
```diff
@@ -15,41 +15,62 @@ const { version } = require('../package.json')
...
+                })
+                .catch(err => {
+                    send('fatal_error', err)
+                    process.exit(1)
+                })
+        }
+    },
+}
 
 client.once('ready', async () => {
```
Comment (nanaian) [inline src/bot.js:65]: All client-related code needs to be done in `start` or the `constructor`. Don't place anything outside the class or it'll be impossible to have two `Bot` instances running at once (they'll reuse eachother's `send`, `client`, etc!)

File: wahtson.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
+const chalk = require('chalk')
+const { Wahtson } = require('./src/bot.js')
+let bot = new Wahtson();
```
Comment (nanaian) [inline wahtson.js:3]: Remove semi.
Prettier isn't running on files outside `src`, edit the args passed to prettier in `package.json` and `.github/prettier.yml` to fix that please.

File: src/config.js [hunk 12]
```diff
@@ -3,36 +3,33 @@ const path = require('path')
...
 
 module.exports = {
-    async load() {
+    async load(CONFIG_TOML_PATH) {
+        const { send } = require('./bot.js')
         const source = await p(fs.readFile)(CONFIG_TOML_PATH).catch(async err => {
-            console.log(chalk.red('config.toml not found! copying the example file...'))
+            send('warning', 'config.toml not found! copying the example file...')
             await p(fs.copyFile)(CONFIG_EXAMPLE_PATH, CONFIG_TOML_PATH)
             return await p(fs.readFile)(CONFIG_TOML_PATH)
```
Comment (nanaian) [inline src/config.js:18]: This behaviour should be removed; instead throw an error that can be caught by `wahtson.js` which does this.
We shouldn't mess with the filesystem or console.

File: src/config.js [hunk 13]
```diff
@@ -56,9 +52,8 @@ module.exports = {
             isOk = await testFn(cache[key])
         } finally {
             if (!isOk) {
-                console.error(chalk.red(`config.toml '${key}' is invalid`))
+                send('config_error', `config.toml '${key}' is invalid`)
 
-                await open(CONFIG_TOML_PATH, { app: 'notepad', wait: true })
                 return await this.get(key, testFn)
```
Comment (nanaian) [inline src/config.js:57]: (Not part of this PR, but just `return` is all that's required here..)

File: wahtson.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+bot.start();
+
+bot.on('info', (event) => {
+    process.stdout.write('\n')
+    process.stdout.write(chalk[event[1]](event[0]))
+})
+bot.on('action', (event) => {
+    process.stdout.write(`${(event[3] == true) ? ' | ' : '\n' }`)
+    process.stdout.write(chalk.grey(`${event[0]}. ${event[1].type}`))
```
Comment (nanaian) [inline wahtson.js:16]: Much better to use an object for `event` rather than an array with arbitrary indexes.

File: wahtson.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    if(event[2]) {
+        process.stdout.write(chalk.magenta(' skipped'))
```
Comment (nanaian) [inline wahtson.js:18]: These should probably just be one big `console.log`? I know I used `process.stdout.write` originally but here you're not flushing `\n` to the console at any point.

File: wahtson.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+let bot = new Wahtson();
+
+process.title = `WAHtson ${bot.version}`
+console.log(`WAHtson ${bot.version}`)
+
+bot.start();
+
+bot.on('info', (event) => {
+    process.stdout.write('\n')
+    process.stdout.write(chalk[event[1]](event[0]))
```
Comment (nanaian) [inline wahtson.js:12]: Would rather that we just used a constant colour for info. If we want more then use a different event. The API here should be consistent with `warning`

File: wahtson.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    if(event[2]) {
+        process.stdout.write(chalk.magenta(' skipped'))
+    }
+})
+bot.on('warning', (event) => {
+    console.log(chalk.orange(event))
+})
+bot.on('error', (event) => {
+    console.log(chalk.red(event))
+})
```
Comment (nanaian) [inline wahtson.js:26]: `error` typically implies that `event` will have the type `Error`.... but they're string logs here.
Perhaps it would be better to just have a `log` event which is passed `{ message: string, type: number }`? The `type` could be one of `const DEBUG = 0, INFO = 1, WARN = 2, ERROR = 3` etc. Then if you wanted to silence, say, all logs less important than `WARN` you can just `if (type >= WARN)`.

File: src/bot.js (hunk 6 above)
Comment (barnabwhy) [reply src/bot.js:111]: rev encountered a thing where because each action was on a new line, his terminal got flooded with complex stuff.
so i added a `limit_log` option onto any event then the actions that just allows the event reciever know that they don't want that to spam in the console.

tl;dr: It prints as one line instead of multiple if `limit_log` is true

File: src/bot.js (hunk 9 above)
Comment (barnabwhy) [reply src/bot.js:31]: I tried but it gave an error, i also needed to use it from `config.js` so it was easier to put it on module.exports

File: src/bot.js (hunk 10 above)
Comment (nanaian) [reply src/bot.js:71]: See `send`.
